### PR TITLE
Add extractPrefix to IPPrefix and IPSet

### PR DESCRIPTION
**Pull request by majst01**
Hi,

Thanks for this library, i enjoy using it as a more convenient api compared to the stdlib variant.
I have a use case where i need to extract a prefix with given bit length from a existing larger prefix/set.

If you find this addition useful, please consider it.

### CODE REVIEW

**Comment by bradfitz** [comment]
I'm not exactly clear on what this is supposed to do based on its docs. Is this the normal name of this operation? Can you point to docs online about it?

As for implementation, though: the repeated calls to `.Range()` and `.Prefixes()` aren't cheap. You should pull those out into variables and minimize how often you need to call those methods.

/cc @mdlayher too.

**Comment by majst01** [comment]
Hi @bradfitz,

tried to clarify the usage and unexported the lesser useful implementation.

I am the author of https://github.com/metal-stack/go-ipam which is the IPAM backend implementation we use for https://github.com/metal-stack which is a bare metal server provisioning.

In the current implementation of go-ipam we dont support ipv6 and variable child prefix length. I added ipv6 support by replacing our bad ipv4 math with the netaddr math. And for the child prefix calculation i use your IPSet.Prefixes function to divide a given Prefix in smaller chunks: https://github.com/metal-stack/go-ipam/pull/36

We use these child prefixes to give every tenant its own address space for his servers, and this address space is part of a existing bigger network.

Hope this helps and thank for looking at my PR.

**Comment by mdlayher** [comment]
I don't have strong feelings about this at the moment as I haven't made any use of the IPSet type in my own code.

I've previously used https://github.com/mikioh/ipaddr for some of my IPAM-like applications, but my netaddr code these days really just cares about /64s.

**Comment by danderson** [comment]
I'm on the fence about this. It means that netaddr starts to get into the IPAM business, which is not what IPSet was originally made for. Then again, prefix extraction is a bunch of subtle code that is annoying to write, so having a single well-debugged implementation seems nice. If I were still developing MetalLB, I would probably use this netaddr method.

Reviewing on the tentative assumption that we'll accept the new API.

File: netaddr.go [hunk 1]
```diff
@@ -1086,6 +1086,78 @@ func (s *IPSet) RemoveSet(b *IPSet) {
...
+// childPrefix, _ := set.ExtractPrefix(18)
+//
+// childPrefix will be: "192.168.0.0/18"
+func (s *IPSet) ExtractPrefix(length uint8) (IPPrefix, error) {
+	prefixes := s.Prefixes()
+	if len(prefixes) == 0 {
+		return IPPrefix{}, fmt.Errorf("no more child prefixes contained in prefix pool")
+	}
+	existingPrefixes := make(map[uint8]IPPrefix)
+	for _, prefix := range prefixes {
```
Comment (danderson) [inline netaddr.go:1130]: I think you can make this more efficient:

```
var bestFit IPPrefix
for _, prefix := range prefixes {
  if prefix.Bits > length {
    continue
  }
  if bestFit.IP.IsZero() || prefix.Bits > bestFit.Bits {
    bestFit = prefix
  }
  if bestFit.Bits == length {
    // exact match, done
    break
  }
}
```

At the end of that iteration, bestFit will be the best prefix to use - either an exact match, or the closest in size - or bestFit.IP.IsZero()==true if no suitable prefix exists. As a bonus, it exits early if it finds an exact match, and you can remove the explicit test for len()==0 above.

**Review by danderson — COMMENTED** [review]
Looks good in general, but I have suggestions for how to rewrite it to be more efficient.

An open question I have: IPAMs generally offer multiple IP allocation policies, and people who use IPAMs tend to care deeply about these things. This implementation implements one specific allocation strategy (best fit, with no preference for how top-level prefixes get consumed), which I would summarize as "don't care, I just want an IPPrefix".

Is it right for us to give that strategy priority over all others? Strictly speaking it's not a huge deal, because IPSet exposes enough primitives to let people implement other strategies externally, but then why are we giving one specific strategy prominent placement in netaddr?

"Because it's a good implementation for people who don't care about IPAM allocation strategies" is probably an okay answer, but I'm curious what others think.

File: netaddr.go [hunk 2]
```diff
@@ -1086,6 +1086,78 @@ func (s *IPSet) RemoveSet(b *IPSet) {
 	}
 }
 
+// extractPrefix will return a Prefix with a bitlength of length which is part of the given Prefix.
+func (p IPPrefix) extractPrefix(length uint8) (IPPrefix, error) {
```
[the remark this answers is not in the document]
Comment (majst01) [reply netaddr.go:1090]: Ha, right, done

File: netaddr.go [hunk 3]
```diff
@@ -1086,6 +1086,78 @@ func (s *IPSet) RemoveSet(b *IPSet) {
...
+	}
+	for _, srp := range srpfx {
+		if srp.Bits == length {
+			return srp, nil
+		}
+	}
+	return IPPrefix{}, fmt.Errorf("no prefix with length:%d found in %s", length, p)
+}
+
+// ExtractPrefix will return a Prefix with a bitlength of length which is part of the given IPSet.
```
[the remark this answers is not in the document]
Comment (majst01) [reply netaddr.go]: done

File: netaddr.go [hunk 4]
```diff
@@ -1086,6 +1086,78 @@ func (s *IPSet) RemoveSet(b *IPSet) {
...
+// The Prefix is also removed from the IPSet.
+//
+// Unlike RemovePrefix this function will not require you to give a Prefix, instead you
```
[the remark this answers is not in the document]
Comment (majst01) [reply netaddr.go]: done as well

File: netaddr.go [hunk 5]
```diff
@@ -1086,6 +1086,78 @@ func (s *IPSet) RemoveSet(b *IPSet) {
...
+// Example usage:
+// pfx,_ := netaddr.ParsePrefix("192.168.0.0/16")
+// set.AddPrefix(pfx)
+// childPrefix, _ := set.ExtractPrefix(18)
+//
+// childPrefix will be: "192.168.0.0/18"
+func (s *IPSet) ExtractPrefix(length uint8) (IPPrefix, error) {
+	prefixes := s.Prefixes()
+	if len(prefixes) == 0 {
+		return IPPrefix{}, fmt.Errorf("no more child prefixes contained in prefix pool")
```
[the remark this answers is not in the document]
Comment (majst01) [reply netaddr.go]: also

File: netaddr.go (hunk 1 above)
Comment (majst01) [reply netaddr.go:1105]: Unsure, don't understand really, bestFit is not set in you example code.

**Comment by bradfitz** [comment]
Naming-wise, maybe `RemoveFreePrefix(bitLen int) (p IPPrefix, ok bool)` ? Or `RemovedSomePrefixOfLength`, etc.

Then it's obvious it removes it, and it matches the other `RemoveXXX` methods.

File: netaddr.go (hunk 1 above)
Comment (danderson) [reply netaddr.go:1105]: It's set in the loop, in the second if. It starts out as the zero value, then gets overridden by the first "big enough" prefix, and gets updated as more precise prefixes are found.

**Comment by majst01** [comment]
Renamed to what @bradfitz suggested, inlined simple func, force pushed to one commit.

**Comment by mdlayher** [comment]
Looks like this needs a rebase.

File: netaddr.go [hunk 6]
```diff
@@ -1132,6 +1132,39 @@ func (s *IPSet) AddRange(r IPRange) {
 	s.in = append(s.in, r)
 }
 
+// RemoveFreePrefix removes and returns a Prefix of length bits from the IPSet.
+// The Prefix is also removed from the IPSet.
```
Comment (bradfitz) [inline netaddr.go:1136]: this sentence seems redundant with the two "remove" words on the previous line (which, while redundant, are conventional)

Comment (majst01) [reply netaddr.go]: done and rebased

**Review by danderson — APPROVED** [review]

**Comment by majst01** [comment]
Thanks guys !

**Pull request merged**